### ISSUE TITLE
Idea: Encourage more local variables

### ISSUE THREAD

**Issue by mcandre** [koalaman/shellcheck#1954]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
foo() {
    x=42
    echo "$x"
}
```

#### Here's what shellcheck currently says:

(nothing)

#### Here's what I wanted or expected to see:

*We noticed that you appear to be using a modern bash descendant shell with an unexported, non-command-prefix, function-scoped variable in lowercase outside of the conventional bindings `http_proxy`, `HTTP_PROXY`, `IFS`, `BASH_VERSION`, etc. etc. and think it would be prudent to mark this variable `local` to reduce the likelihood of collision with similar variables from other shell scripts.*

**Comment by christophgil** [comment]
I strongly vote for this.
I keep forgetting the local.

I suggest that global scope variables could be recognized by upper case characters 
or maybe by  leading underscore as in python

Any other variable in functions should be assumed to be local.

Many thanks. Really cool code checker!
Cheers
C

**Comment by brother** [comment]
Upper case variables indicate that the scope is environment global, they are set outside the script. Not sure if I have invented this myself or inherited from somewhere.

**Comment by dgutson** [comment]
I think that this could be a good rule: suggest the usage of `local`
CC @itaranto

**Comment by mirekfranc** [comment]
I must say I was surprised that shellcheck doesn't already do this.  Were there any attempts at implementing it?

**Comment by mirekfranc** [comment]
This is not only problem of explicit assignments.  For example this...

```bash
#!/usr/bin/bash
f()
{
        for i in hello
        do
                echo $i
        done
}

f
echo $i
```

will print "hello" twice.

```bash
$ ./hello
hello
hello
```

**Comment by hramrach** [comment]
There is the additional problem that 'local' is not part of POSIX and some shells refuse to implement it.

Consequently, scoping your variables is non-portable.

It is nice thing to have for shells that implement it but probably should not be the default.

**Comment by dgutson** [comment]
> There is the additional problem that 'local' is not part of POSIX and some shells refuse to implement it.
> 
> Consequently, scoping your variables is non-portable.

shellcheck is already capable of distinguishing the shell implementation, by reading the shebang, thus emitting specific warnings.

**Comment by mirekfranc** [comment]
> It is nice thing to have for shells that implement it but probably should not be the default.

I think that `shellcheck` has some notion of what is checking anyway.  Either with `-s` option, shebang or file extention.

> Consequently, scoping your variables is non-portable.

Not only because of `local` or not.  For example some shells are using `typeset` instead of `local` like `ksh93`, and for some, I think `bash`, these two are aliases.  On top of that, `typeset` in `ksh93` uses static scoping and  works only in `function function {}` style definition, not in `function () {}` style.

Consider the following:

```bash
x=7
function foo { echo "in foo: $x"; }
function bar { typeset x=42; foo; }
bar
foo
```

It prints...
```
in foo: 42
in foo: 7
```
... in bash, but in ksh93, it prints...
```
in foo: 7
in foo: 7
```
... because of static scoping.

Now consider, the same program, but the other style of function definition...

```
x=7
foo() { echo "in foo: $x"; }
bar() { typeset x=42; foo; }
bar
foo
```

The bash gives the same output, but `ksh93` now prints...

```
in foo: 42
in foo: 42
```
... because typeset is ignored.

I'm not sure how other ksh implementations behave and I do realize, it's a minefield, but that deoesn't detract from potential usefulness of some form of warning.  My only regret is that my haskell is extremely rusty after some 19 years, so I'm unlikely to contribute myself to this.
